### ISSUE TITLE
EntityAssigner not using proper EntityManager

### ISSUE THREAD

**Issue by emrosenf** [mikro-orm/mikro-orm#246]
I am using `ApolloServer` and passing a forked `em` into the `ctx: Context`. I am watching sql queries with debug mode on and notice that for a very simple case when fetching relationships there are fewer queries than expected, meaning that somehow my forked `em` is being polluted with global context from previous requests.

I am still familiarizing myself with the code, but I believe the error might stem from [EntityAssigner L15](https://github.com/mikro-orm/mikro-orm/blob/master/lib/entity/EntityAssigner.ts#L15)

```typescript
  static assign<T extends AnyEntity<T>>(entity: T, data: EntityData<T>, onlyProperties: AssignOptions | boolean = false): T {
    const em = wrap(entity).__em;
    ...
```

which is called by [EntityManager L222](https://github.com/mikro-orm/mikro-orm/blob/d4eddfd0e3b0e082829ca8c63669e88b13ab510c/lib/EntityManager.ts#L222)

```typescript
    // add to IM immediately - needed for self-references that can be part of `data` (and do not trigger cascade merge)
    this.getUnitOfWork().merge(entity, [entity]);
    EntityAssigner.assign(entity, data as EntityData<T>, true);
    this.getUnitOfWork().merge(entity); // add to IM again so we have correct payload saved to change set computation
```

`EntityAssigner.assign` is accessing a different instance of `em` that has more objects defined in its `unitOfWork.identityMap` than the forked instance of `em` that we had been using in `EntityManager.merge` previously in the call stack.

Here is the relevant portion of my `index.ts` just to show you that I am following all of your suggestions and still having this issue:

```typescript
  app.use((req, res, next) => {
    RequestContext.create(orm.em, next);
  });

  const apolloServer = new ApolloServer({
    schema: await buildSchema({
      resolvers: [UserResolver]
    }),
    context: ({ req, res }) => {
      const em = orm.em.fork()
      em.clear()
      // confirming that each request starts with a clean map
      console.log(em.getUnitOfWork().getIdentityMap())
    
      return {
        req,
        res,
        em,
      }
    }
  });
```

**Comment by B4nan** [comment]
If you are creating the request context manually (via the `em.fork()`, you do not need to use the `RequestContext` helper (on the app level bootstrap, you should use only one of those). The only downside is that you will need to use the forked `em` all the time in your requests manually (aka this breaks DI on controller level, you can't inject EM or repositories and use those as they would be associated with wrong EM - and that is what the `RequestContext` is for). Or on the other hand you could use it and drop the manual forking. 

Also note that `em.getUnitOfWork()` will return UoW stored in current request context (the one started from `RequestContext` helper). So in your snippet, you are creating the fork, but `console.log(em.getUnitOfWork().getIdentityMap())` will return the UoW (and its IM) from the one created from `RequestContext` helper. 

Same problem applies to `em.clear()`, under the hood it will also prefer the UoW from `RequestContext`. So in your snippet, you create the fork, but you clear different EM than you think. But the `em.fork()` itself will clear the IM by default, so you can just omit the line.

If you use latest v3, there is second parameter in `em.fork()` that allows you to specify whether the new (forked) EM will use its own UoW/IM, or it will fallback to the one created `RequestContext`. By default `em.fork()` will give you clean IM and the new EM will not fallback to the `RequestContext`.

Btw as part of `em.merge(), the entity is also reassigned with internal EM reference. That is the one you get via `wrap(entity).__em`. 

> EntityAssigner.assign is accessing a different instance of em that has more objects defined in its unitOfWork.identityMap than the forked instance of em that we had been using in EntityManager.merge previously in the call stack.

If this is really true then it sounds like you are loading the particular entity from some global context, and not from the request handler. Could you share more information about that? What code, and what entity (created/loaded where) causes this?

But as I said, `em.merge` will reassign the internal `em` reference so it sounds really weird that you would get a wrong one in the following line.

**Comment by emrosenf** [comment]
[Here's a minimal repo](https://github.com/emrosenf/mikro-orm-bug/tree/master) to reproduce. I appreciate you taking a look.

In this repro I only use `em.fork()`. You can see I commented out the `RequestContext` middleware. In my testing, the `domain.run` calls `next()` but returns before my `ApolloServer` request is run, which ends the domain lifecycle, making it useless.

**Comment by B4nan** [comment]
Thanks for the repro, that will help a lot! Will try to look into this soon.

**Comment by emrosenf** [comment]
Much appreciated. This library looks super promising, I really appreciate your hard work.

**Comment by emrosenf** [comment]
I did a bit more digging and here is what I believe is happening:

The getter for `wrap(entity).__em` is created by `EntityHelper.decorate`, which is called by `metadata.decorate`, which only happens in the bootstrap phase of `MikroORM.init`.

Therefore, in future requests, even though a forked `em` is passed, the `wrap(entity).__em` returns the "global" one created in the bootstrap phase by `MikroORM.init`.

I've also found that I have the desired outcome if every request I fork the `em` and clear the global `em` like so:

```typescript
  const apolloServer = new ApolloServer({
    schema: await buildSchema({
      resolvers: [UserResolver]
    }),
    context: ({ req, res }) => {
      const em = orm.em.fork()
      orm.em.clear()

      return {
        req,
        res,
        em,
      }
    }
  });
```

This also supports my conclusion that the global `em` is being used. Hope that's helpful. In case you already pulled my repo, I added a few more comments to make it super easy to trace.

**Comment by B4nan** [comment]
> Therefore, in future requests, even though a forked em is passed, the wrap(entity).__em returns the "global" one created in the bootstrap phase by MikroORM.init.

As I said, `em.merge()` does replace this reference with the current one, so this should not be the problem. 

> In this repro I only use em.fork(). You can see I commented out the RequestContext middleware. In my testing, the domain.run calls next() but returns before my ApolloServer request is run, which ends the domain lifecycle, making it useless.

If you do not use `RequestContext` helper, then there is no magic involved (no domain api used).

**Comment by emrosenf** [comment]
I'm sorry, I think I was unclear and made this thread confusing. I am not using `RequestContext` at all, although I had previously (and unsuccessfully) tried it. I am only using `em.fork()`.

Please do check out my repo when you have a chance, I really believe there is unexpected behavior. Here is why I believe `merge` is not working as you say.

In my example I am fetching two linked objects: a `User` and a `Tag`. Here is what happens on the second request.

1. The `user` is fetched and `merge` is called. The forked `em`'s `IdentityMap` is empty. `entity === undefined`, and so `__em` is not redefined. Therefore, `user.__em` still points to the global reference, not the forked one.

2. The `EntityAssigner` will merge all properties from `user.__em` (the global reference) into the newly fetched object. This includes the details of the `tag` from the previous request, even though it hasn't been fetched yet in this request.

```typescript
  merge<T extends AnyEntity<T>>(entityName: EntityName<T> | T, data?: EntityData<T> | boolean, refresh?: boolean): T {
    ...
    // See Point 1
    let entity = this.getUnitOfWork().tryGetById<T>(entityName, data as FilterQuery<T>);
    if (entity) {
      Object.defineProperty(entity, '__em', { value: this, writable: true });
    }
    ...
    // See Point 2
    this.getUnitOfWork().merge(entity, [entity]);
    EntityAssigner.assign(entity, data as EntityData<T>, true);
    this.getUnitOfWork().merge(entity); // add to IM again so we have correct payload saved to change set computation

    return entity;
  }
```

**Comment by B4nan** [comment]
Ok so the problem is that the `em.merge()` is actually not correcting the internal `__em` value as it should be (not for the user entity because it was not looked up by primary key). 

Will try to publish new version soon.

**Comment by emrosenf** [comment]
Makes sense. Thanks!

**Comment by B4nan** [comment]
Btw as a workaround, you can do this correction yourself. I think it should be enough to merge it again:

    const user = await ctx.em.findOne(User, {
      firstName: 'user-1',
    });

    if (user) {
      ctx.em.merge(user);
    }

    return user;

If not, you can reassign it directly:

    if (user) {
      wrap(user).__em = ctx.em;
    }

**Comment by B4nan** [comment]
Fixed in `3.0.0-rc.1`.
